File: JavaTraining/src/main/java/Selenium/Common/PageEntities/HtmlSection.java
package Selenium.Common.PageEntities;

import Selenium.Common.Helpers.CustomImplicitTimeout;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class HtmlSection {
    protected final int QUICK_SEARCH_TIMEOUT = 2;
    private Logger log = LogManager.getFormatterLogger(this.getClass().getName());
    private WebDriver webDriver;
    private By by;

    private HtmlSection() {}

    public HtmlSection(WebDriver webDriver, By by) {
        this.webDriver = webDriver;
        this.by = by;
    }

    public WebElement WebElement() {
        return by != null ? webDriver.findElement(by) : webDriver.findElement(By.xpath(".//"));
    }

    public String Text() {
        try (var ignored = new CustomImplicitTimeout(webDriver, QUICK_SEARCH_TIMEOUT)) {
            return WebElement().getText();
        } catch (Exception ex) {
            log.error("Error has occurred during an attempt to get the text of the element! Message: %s", ex.getMessage());
            return "";
        }
    }

    public Boolean IsPresent() {
        try (var ignored = new CustomImplicitTimeout(webDriver, QUICK_SEARCH_TIMEOUT)) {
            webDriver.findElement(by);
            return true;
        } catch (Exception ex) {
            log.warn("The element hasn't been found by '$s' in the DOM!", by.toString());
            return false;
        }
    }

    public Boolean IsAbsent() {
        try (var ignored = new CustomImplicitTimeout(webDriver, QUICK_SEARCH_TIMEOUT)) {
            return webDriver.findElements(by).size() == 0;
        } catch (Exception ex) {
            log.error("Error has occurred during an attempt to get the element! Message: %s", ex.getMessage());
        }
        return false;
    }

    public Boolean IsDisplayed() {
        return WebElement().isDisplayed();
    }

    public Boolean IsHidden() {
        return !WebElement().isDisplayed();
    }
}

File: C#/SeleniumTest/SeleniumTest/Core/HtmlEntities/HtmlSection.cs
using log4net;
using OpenQA.Selenium;
using OpenQA.Selenium.Internal;
using OpenQA.Selenium.Support.UI;
using System;

namespace SeleniumTest.Core
{
    public class HtmlSection
    {
        protected static Configuration config = Configuration.GetInstance();
        protected static int explicitTimeout = int.Parse(config.GetValue("explicitTimeout"));
        protected static int quickSearchTimeout = int.Parse(config.GetValue("quickSearchTimeout"));
        protected ILog log;
        private IWebDriver _webDriver;
        private IWebElement _webElement;
        private By by;

        private HtmlSection()
        {
            log = LogManager.GetLogger(this.GetType());
        }

        public HtmlSection(IWebDriver webDriver, By by) : this()
        {
            this._webDriver = webDriver;
            this.by = by;
        }

        public HtmlSection(IWebElement webElement, By by) : this()
        {
            this._webElement = webElement;
            this.by = by;
        }

        public HtmlSection(IWebElement webElement) : this()
        {
            this._webElement = webElement;
        }

        public IWebElement WebElement
        {
            get
            {
                return by != null && _webElement != null
                   ? _webElement.FindElement(by)
                   : by != null
                       ? _webDriver.FindElement(by)
                       : _webElement;
            }
        }

        public IWebDriver WebDriver => _webDriver != null ? _webDriver : ((IWrapsDriver)_webElement).WrappedDriver;

        protected string InternalId => by != null ? by.ToString() : "by.xpath //";

        public void Click()
        {
            try
            {
                WebElement.Click();
            }
            catch (Exception ex)
            {
                log.Error($"An error has occurred while clicking on the element! Message: {ex.Message}");
            }
        }

        public virtual string Text
        {
            get
            {
                try
                {
                    using (new CustomImplicitTimeout(WebDriver, quickSearchTimeout))
                    {
                        return WebElement.Text;
                    }
                }
                catch (Exception ex)
                {
                    log.Error($"Error has occurred during an attempt to get the text of the element! Message: {ex.Message}");
                    return "";
                }
            }
            set { }
        }

        public bool IsPresent
        {
            get
            {
                try
                {
                    using (new CustomImplicitTimeout(WebDriver, quickSearchTimeout))
                    {
                        var testWebElement = WebElement;
                        return true;
                    }
                }
                catch (Exception)
                {
                    log.Warn($"IsPresent: The element hasn't been found by '{by.ToString()}' locator in the DOM!");
                    return false;
                }
            }
        }

        public bool IsAbsent
        {
            get
            {
                try
                {
                    using (new CustomImplicitTimeout(WebDriver, quickSearchTimeout))
                    {
                        int numberOfElements = _webElement != null
                            ? _webElement.FindElements(by).Count
                            : _webDriver.FindElements(by).Count;

                        return numberOfElements == 0;
                    }
                }
                catch (Exception ex)
                {
                    log.Error($"IsAbsent: Error has occurred during an attempt to get the element! Message: {ex.Message}");
                }
                return false;
            }
        }

        public bool IsDisplayed
        {
            get
            {
                try
                {
                    using (new CustomImplicitTimeout(WebDriver, quickSearchTimeout))
                    {
                        return WebElement.Displayed;
                    }
                }
                catch (NoSuchElementException ex)
                {
                    log.Error($"IsDisplayed: Error has occurred during an attempt to get the element! Message: {ex.Message}");
                }
                return false;
            }
        }

        public bool IsHidden
        {
            get
            {
                try
                {
                    using (new CustomImplicitTimeout(WebDriver, quickSearchTimeout))
                    {
                        return !WebElement.Displayed;
                    }
                }
                catch (NoSuchElementException ex)
                {
                    log.Error($"IsHidden: Error has occurred during an attempt to get the element! Message: {ex.Message}");
                }
                return false;
            }
        }

        public void waitUntil(Func<IWebDriver, bool> condition) => new WebDriverWait(WebDriver, TimeSpan.FromSeconds(explicitTimeout)).Until((WebDriver) => condition.Invoke(WebDriver));
    }
}
